### ISSUE TITLE
Continuing the sign-in flow in a popup

### ISSUE THREAD

**Issue by cbiesinger** [fedidcg/FedCM#555]
(this has been split out of #477 )

FedCM’s account chooser only asks for permission to share the user’s name, email address and profile picture. However, commonly Identity Providers (IDPs) want to ask for additional permissions (e.g. giving access to a user’s calendar, etc.). This is not currently supported in the FedCM API. In other cases, the IDP needs to ask for additional information before returning the token to the relying party (RP), such as requesting re-authentication, verifying up-to-date contact information, etc.

**Comment by cbiesinger** [comment]
## Proposal

We are proposing to extend the ID assertion response to let the IDP specify a field called `continue_on`. This field would contain a (same-origin) URL that the user agent will show in a popup so that the user can continue the signin flow.

Once finished, the identity provider will call a new function `IdentityProvider.resolve` in the popup to indicate that the flow is finished and to provide the token to pass back to the RP.


## IDL


```
dictionary IdentityResolveOptions {
 USVString accountId;
};

partial interface IdentityProvider
    // Allows an IdP to return a token to the RP from the content area,
    // as opposed to over HTTP with the id_assertion_endpoint.
    static Promise<undefined> resolve(USVString token,
        optional IdentityResolveOptions options = {});
};
```


And the dictionary for the ID assertion response would be extended like this:


```
dictionary IdentityProviderToken {
  optional USVString token;
  optional USVString continue_on;
};
```



## Example


```js
{
  // In the id_assertion_endpoint, instead of returning a typical
  // "token" response, the IdP decides that it needs the user to
  // continue on pop-up window:
  "continue_on": "/oauth/authorize?scope=..."
}
```


The popup might then contain code like this:


```js
document.getElementById("allow_btn").addEventListener("click", async () => {
  let accessToken = await fetch(`/generate_access_token.cgi`);
  // Closes the window and resolves the promise (that is still hanging
  // in the relying party's renderer) with the value that is passed.
  IdentityProvider.resolve(accessToken);
});
```


The user agent will then close the popup and resolve the promise (returned from navigator.credentials.get) with the provided token.

If for some reason the user has changed in the popup (e.g. the IDP offers a “switch user” function or in delegation cases), the resolve call takes an optional second argument allowing something like:


```js
  IdentityProvider.resolve(token, {accountId: "1234");
```


If not provided, the user agent will assume the account ID is unchanged from what the user has selected.


## Considerations

Because this happens after the user has selected an account, it is safe to allow access to unpartitioned cookies in the popup. However, we still do not allow direct communication with the RP – the popup behaves as if `noopener,noreferrer` was specified (thus, `window.opener` is null). The only way to communicate with the RP should be a one-time `IdentityProvider.resolve` call, which also closes the popup.

We only want to allow popups when we had a user interaction. Therefore, we allow them when the user selected an account in the account chooser, or in button mode (which always requires user activation). But we don’t want to allow the popup for `mediation: silent` (a popup is pretty loud), and for `mediation: optional` for the widget mode we check for transient user activation. Otherwise, we open no popup and reject the promise.

Due to its versatility, this proposal can also be used for cases where no token will be returned, such as for showing custom error messages. We think this is a feature, not a bug.

**Comment by achimschloss** [comment]
Some thoughts:

- Are there restrictions on re-directs within the pop-up? In the current prototype there aren't, but resolve / close can only be called from the IDP domain - As discussed re-directs are essential for IDPs and must be possible
- Are there requirements for user interaction in the popup? There should not be given the IDP might not need a user interaction to continue, and the browser UX already had a user interaction

**Comment by samuelgoto** [comment]
> Posting this on behalf of @panva

It would be cool if the IdP had a HTTP header version of the `IdentityProvider.resolve()` call to fit within the current deployment.

**Comment by samuelgoto** [comment]
> Also from @panva:

Do the requests shown in a window in response to the "IDP login dialog" flow include the "Sec-Fetch-Dest: webidentity" HTTP header? If not, why?

**Comment by samuelgoto** [comment]
> It would be cool if the IdP had a HTTP header version of the IdentityProvider.resolve() call to fit within the current deployment.

That seems reasonable to me. @cbiesinger, any concern?

**Comment by samuelgoto** [comment]
> Do the requests shown in a window in response to the "IDP login dialog" flow include the "Sec-Fetch-Dest: webidentity" HTTP header? If not, why?

I believe the answer is "no". @cbiesinger ack?

**Comment by vaceksimon** [comment]
+1 to the HTTP header version of the js call. This could also be considered for [the dynamic sign-in flow](https://developers.google.com/privacy-sandbox/blog/fedcm-chrome-116-updates#what_if_the_user_session_expires_let_the_user_sign_in_through_a_dynamic_sign-in_flow). After authenticating the user, the IdP could send the successful authn response only with extra headers causing the pop-up window to close, without the need to modify the signing in process any futher.

**Comment by panva** [comment]
> > It would be cool if the IdP had a HTTP header version of the IdentityProvider.resolve() call to fit within the current deployment.
> 
> That seems reasonable to me. @cbiesinger, any concern?

For clarity I don't think it would be cool per se. I think it is a prerequisite for being able to fit the FedCM signals to an existing IdP login interaction rather than having to define a new login form response that is a blank HTML with a <script> block.

**Comment by cbiesinger** [comment]
> > Also from @panva:
> 
> Do the requests shown in a window in response to the "IDP login dialog" flow include the "Sec-Fetch-Dest: webidentity" HTTP header? If not, why?

They do not. Similar to the IDP login flow, the continuation URL gets shown in a regular browser popup so the destination is not semantically "the FedCM flow", hence we use the regular destination. (if we wanted to use webidentity, that would also prompt the question what we should send for subresources)

**Comment by cbiesinger** [comment]
> > > It would be cool if the IdP had a HTTP header version of the IdentityProvider.resolve() call to fit within the current deployment.
> > 
> > 
> > That seems reasonable to me. @cbiesinger, any concern?
> 
> For clarity I don't think it would be cool per se. I think it is a prerequisite for being able to fit the FedCM signals to an existing IdP login interaction rather than having to define a new login form response that is a blank HTML with a <script> block.

Makes sense to me, we could definitely define an HTTP header for this, similar to how we have an HTTP version of the login status API.

**Comment by panva** [comment]
FWIW the comments @samuelgoto re-posted from our private conversation don't parcularly pertain to continuation flow but rather any window popped up by FedCM.

**Comment by cbiesinger** [comment]
@panva so you are suggesting that you would also like an HTTP version of IdentityProvider.close?

**Comment by panva** [comment]
> @panva so you are suggesting that you would also like an HTTP version of IdentityProvider.close?

Yes, that's the one my message to @samuelgoto wss actually about.

**Comment by panva** [comment]
Anything I can't do with headers as an IdP is an annoyance because it means i need to add intermediary script only pages in my existing flow.

**Comment by ThisIsMissEm** [comment]
My only real problem with this, and why I think #590 is better is because this forces popups to be used, rather than just allowing the developer to choose what to do, e.g., full-page redirect or popup.

**Comment by cbiesinger** [comment]
The problem with redirects is that it does not match super well to a JavaScript API, because by definition it can't return anything to the JavaScript because the document is gone.

**Comment by cbiesinger** [comment]
As a replacement for IdentityProvider.resolve? Yes, that was also requested elsewhere, we can definitely do that
